server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java

```diff
@@ -103,7 +104,7 @@ public final Query toQuery(QueryShardContext context) throws IOException {
             if (boost != DEFAULT_BOOST) {
                 if (query instanceof SpanQuery) {
                     query = new SpanBoostQuery((SpanQuery) query, boost);
-                } else {
```

Comment:
For my knowledge, I'm wondering why this change (and the ones to `AbstractQueryTestCase`) were necessary.

Reply:
I made this change as a way to keep tests simple. For instance here is what `ConstantScoreQueryBuilderTests#doAssertLuceneQuery` looks like in master today.

```
    @Override
    protected void doAssertLuceneQuery(ConstantScoreQueryBuilder queryBuilder, Query query, QueryShardContext context) throws IOException {
        Query innerQuery = queryBuilder.innerQuery().toQuery(context);
        if (innerQuery == null) {
            assertThat(query, nullValue());
        } else {
            assertThat(query, instanceOf(ConstantScoreQuery.class));
            ConstantScoreQuery constantScoreQuery = (ConstantScoreQuery) query;
            assertThat(constantScoreQuery.getQuery(), instanceOf(innerQuery.getClass()));
        }
    }
```

This test only worked because most queries on unmapped fields would create the same query as on a `keyword` field. But with this change, queries on unmapped fields get rewritten as a `MatchNoneQueryBuilder`. And when its inner query rewrites to a `MatchNoneQueryBuilder`, `ConstantScoreQueryBuilder` itself rewrites to a `MatchNoneQueryBuilder`. So I updated the logic this way:

```
    @Override
    protected void doAssertLuceneQuery(ConstantScoreQueryBuilder queryBuilder, Query query, QueryShardContext context) throws IOException {
        Query innerQuery = queryBuilder.innerQuery().rewrite(context).toQuery(context);
        if (innerQuery == null) {
            assertThat(query, nullValue());
        } else if (innerQuery instanceof MatchNoDocsQuery) {
            assertThat(query, instanceOf(MatchNoDocsQuery.class));
        } else {
            assertThat(query, instanceOf(ConstantScoreQuery.class));
            ConstantScoreQuery constantScoreQuery = (ConstantScoreQuery) query;
            assertThat(constantScoreQuery.getQuery(), instanceOf(innerQuery.getClass()));
        }
    }
```

But this fails if the inner query is wrapped in a `BoostQuery`. So I had to either change `AbstractQueryBuilder` to no longer wrap `MatchNoDocsQuery` with a boost, or change this test (and a couple other ones IIRC) to check whether the inner query might be wrapped inside a `BoostQuery`. I chose the former, which sounded simpler to me.

Reply:
Thanks for the clear explanation!